walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,101 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        /*
+            params:
+                MeetingRequest request: specifies (1) a duration of time and (2) a collection of attendees
+                Collection<Event> events: specifies the times during day which users are busy
+
+            RETURN: A collection of intervals (time ranges) which are valid solutions to the meeting request. Valid 
+            meaning each interval is at least the specified duration and does not overlap with any events.
+        */
+
+        // Brief Approach: Add intervals starting from time 0 going forward through the sorted events by start time.
+
+        ArrayList<TimeRange> validIntervals = new ArrayList<TimeRange>();
+
+
+        /*          Edge Cases            */
+
+        // Request duration must be less than or equal to a day
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+            // Return an empty collection of valid intervals since non should exist
+            return validIntervals;
+        }
+
+        if (request.getAttendees().isEmpty() || events.isEmpty()){
+            // The whole day could then be scheduled
+            validIntervals.add(TimeRange.WHOLE_DAY);
+            return validIntervals;
+        }
+
+        // Sort the events by increasing start time
+        ArrayList<Event> sortedEvents = new ArrayList<>(events);      
+
+        Collections.sort(sortedEvents, new Comparator<Event>() {
+            @Override
+            public int compare(Event a, Event b) {
+                return Long.compare(a.getWhen().start(), b.getWhen().start());
+            }
+        });
+        
+        
+        // Greedily loop and take away intervals, capturing an end of an event with vacancy thereafter
+
+        int curr_start = TimeRange.START_OF_DAY;
+        while (!sortedEvents.isEmpty()){
+            Event leastMost = sortedEvents.get(0);
+
+            // Void this event if it does not include members from the meeting request
+            if (!includesRequestMembers(leastMost, request)) {
+                sortedEvents.remove(0);
```

Comment:
`Arraylist.remove()` has a worst case time complexity of O(n) and this case would actually always be O(n). This is because after the first element is removed, the remaining elements would need to be shifted by one. You might want to try using a [`TreeSet`](https://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html) which has O(1) removal and is also presorted.